### DISCUSSION TITLE
how to call nv_read with a custom NvIndexHandle

### DISCUSSION THREAD

**Discussion by CedricLemay** [parallaxsecond/rust-tss-esapi#401] [Q&A] [answered]
The example only shows how to call nv_read after getting the NvIndexHandle from nv_define_space. In my situation, I need to read nv memory that has already been written in the TPM. I am struggling to find the syntax to have my NvIndexHandle set to a custom address. I know it is possible because I can read it using tpm2_tools. Also, it was really easy to do with the old library: https://gitlab.com/starlab-io/tss-sapi/-/blob/master/examples/nvread.rs

Thanks for the help

**Comment by Superhepper** [comment, marked as answer]
I suggest you look in the abstraction module.

https://docs.rs/tss-esapi/latest/tss_esapi/abstraction/nv/fn.read_full.html

To do this by hand can be a little bit complicated.  You would need to retrieve an object handle for the NV memory area by calling ```tr_from_tpm_public``` and convert that into a ```NvIndexHandle``` and then use that handle to read the nv memory area.

**Reply by CedricLemay** [reply]
Thanks you so much, it worked with tr_from_tpm_public.
Here is how I did it:
`    let nv_index_tpm_handle: NvIndexTpmHandle = NvIndexTpmHandle::new(0x1000001).expect("Failed to create NvIndexTpmHandle");
    let obj_handle_from_tpm = context.tr_from_tpm_public(TpmHandle::NvIndex(nv_index_tpm_handle)).expect("Failed to create ObjectHandle from NvIndexTpmHandle");
    let nv_read_result = context.nv_read(NvAuth::Owner, NvIndexHandle::from(obj_handle_from_tpm), 32, 0);`
